# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [182]:
import pandas as pd

In [183]:
df = pd.read_csv("../data/2019_Yellow_Taxi_Trip_Data.csv")
df.head(5)

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5    1.0      0.5   
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [184]:
df.shape

(10000, 18)

#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [185]:
df.columns

Index(['vendorid', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'ratecodeid', 'store_and_fwd_flag',
       'pulocationid', 'dolocationid', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge'],
      dtype='object')

In [186]:
df[['fare_amount', 'tip_amount', 'tolls_amount', 'total_amount']].describe() ### double [[]] if two or more

fare_amount    tip_amount  tolls_amount  total_amount
count  10000.000000  10000.000000  10000.000000  10000.000000
mean      15.106313      2.634494      0.623447     22.564659
std       13.954762      3.409800      6.437507     19.209255
min      -52.000000      0.000000     -6.120000    -65.920000
25%        7.000000      0.000000      0.000000     12.375000
50%       10.000000      2.000000      0.000000     16.300000
75%       16.000000      3.250000      0.000000     22.880000
max      176.000000     43.000000    612.000000    671.800000

In [187]:
df[['fare_amount', 'tip_amount', 'tolls_amount', 'total_amount']].describe().T

count       mean        std    min     25%   50%    75%    max
fare_amount   10000.0  15.106313  13.954762 -52.00   7.000  10.0  16.00  176.0
tip_amount    10000.0   2.634494   3.409800   0.00   0.000   2.0   3.25   43.0
tolls_amount  10000.0   0.623447   6.437507  -6.12   0.000   0.0   0.00  612.0
total_amount  10000.0  22.564659  19.209255 -65.92  12.375  16.3  22.88  671.8

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [188]:

df.trip_distance.max(), max(df.trip_distance)

(38.11, 38.11)

In [189]:
df[df.trip_distance==max(df.trip_distance)][['fare_amount','tip_amount','tolls_amount','total_amount']]

fare_amount  tip_amount  tolls_amount  total_amount
8338        176.0       18.29          6.12        201.21

---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [190]:
import pandas as pd

In [198]:
meteorites = pd.read_csv("../data/Meteorite_Landings.csv")
meteorites.head(5)

name   id nametype     recclass  mass (g)  fall  \
0    Aachen    1    Valid           L5      21.0  Fell   
1    Aarhus    2    Valid           H6     720.0  Fell   
2      Abee    6    Valid          EH4  107000.0  Fell   
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell   
4   Achiras  370    Valid           L6     780.0  Fell   

                     year    reclat    reclong           GeoLocation  
0  01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)

In [192]:
meteorites.rename(columns={'mass (g)':"mass"}, inplace=True)
meteorites.drop(columns=['reclat','reclong'], inplace=True)
meteorites.sort_values(by="mass", ascending=False, inplace=True)
meteorites.head()

name     id nametype      recclass        mass   fall  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found   

                         year             GeoLocation  
16392  01/01/1920 12:00:00 AM   (-19.58333, 17.91667)  
5373   01/01/1818 12:00:00 AM   (76.13333, -64.93333)  
5365   12/22/1575 12:00:00 AM  (-27.46667, -60.58333)  
5370   01/01/1891 12:00:00 AM     (35.05, -111.03333)  
3455   01/01/1898 12:00:00 AM            (47.0, 88.0)

#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [193]:
import pandas as pd
import numpy as np
import datetime

In [194]:
meteorites = pd.read_csv("../data/Meteorite_Landings.csv")
meteorites.head(5)

name   id nametype     recclass  mass (g)  fall  \
0    Aachen    1    Valid           L5      21.0  Fell   
1    Aarhus    2    Valid           H6     720.0  Fell   
2      Abee    6    Valid          EH4  107000.0  Fell   
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell   
4   Achiras  370    Valid           L6     780.0  Fell   

                     year    reclat    reclong           GeoLocation  
0  01/01/1880 12:00:00 AM  50.77500    6.08333     (50.775, 6.08333)  
1  01/01/1951 12:00:00 AM  56.18333   10.23333  (56.18333, 10.23333)  
2  01/01/1952 12:00:00 AM  54.21667 -113.00000    (54.21667, -113.0)  
3  01/01/1976 12:00:00 AM  16.88333  -99.90000     (16.88333, -99.9)  
4  01/01/1902 12:00:00 AM -33.16667  -64.95000   (-33.16667, -64.95)

In [195]:
meteorites['year'] = pd.to_datetime(meteorites['year'],errors = 'coerce' )  # change the format to date-time before
meteorites['only_year'] = pd.DatetimeIndex(meteorites['year']).year
meteorites.head()

name   id nametype     recclass  mass (g)  fall       year    reclat  \
0    Aachen    1    Valid           L5      21.0  Fell 1880-01-01  50.77500   
1    Aarhus    2    Valid           H6     720.0  Fell 1951-01-01  56.18333   
2      Abee    6    Valid          EH4  107000.0  Fell 1952-01-01  54.21667   
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell 1976-01-01  16.88333   
4   Achiras  370    Valid           L6     780.0  Fell 1902-01-01 -33.16667   

     reclong           GeoLocation  only_year  
0    6.08333     (50.775, 6.08333)     1880.0  
1   10.23333  (56.18333, 10.23333)     1951.0  
2 -113.00000    (54.21667, -113.0)     1952.0  
3  -99.90000     (16.88333, -99.9)     1976.0  
4  -64.95000   (-33.16667, -64.95)     1902.0

In [196]:
meteorites.loc[meteorites['only_year'] <= 1970, 'Year above or equel 1970?'] = 'True'
meteorites.loc[meteorites['only_year'] > 1970, 'Year above or equel 1970?'] = 'False'

meteorites.head(16)

name   id nametype      recclass  mass (g)  fall  \
0                Aachen    1    Valid            L5      21.0  Fell   
1                Aarhus    2    Valid            H6     720.0  Fell   
2                  Abee    6    Valid           EH4  107000.0  Fell   
3              Acapulco   10    Valid   Acapulcoite    1914.0  Fell   
4               Achiras  370    Valid            L6     780.0  Fell   
5              Adhi Kot  379    Valid           EH4    4239.0  Fell   
6   Adzhi-Bogdo (stone)  390    Valid         LL3-6     910.0  Fell   
7                  Agen  392    Valid            H5   30000.0  Fell   
8                Aguada  398    Valid            L6    1620.0  Fell   
9         Aguila Blanca  417    Valid             L    1440.0  Fell   
10     Aioun el Atrouss  423    Valid  Diogenite-pm    1000.0  Fell   
11                  Aïr  424    Valid            L6   24000.0  Fell   
12      Aire-sur-la-Lys  425    Valid       Unknown       NaN  Fell   
13                Akaba  426    Valid            L6     779.0  Fell   
14             Akbarpur  427    Valid            H4    1800.0  Fell   
15              Akwanga  432    Valid             H    3000.0  Fell   

         year    reclat    reclong           GeoLocation  only_year  \
0  1880-01-01  50.77500    6.08333     (50.775, 6.08333)     1880.0   
1  1951-01-01  56.18333   10.23333  (56.18333, 10.23333)     1951.0   
2  1952-01-01  54.21667 -113.00000    (54.21667, -113.0)     1952.0   
3  1976-01-01  16.88333  -99.90000     (16.88333, -99.9)     1976.0   
4  1902-01-01 -33.16667  -64.95000   (-33.16667, -64.95)     1902.0   
5  1919-01-01  32.10000   71.80000          (32.1, 71.8)     1919.0   
6  1949-01-01  44.83333   95.16667  (44.83333, 95.16667)     1949.0   
7  1814-01-01  44.21667    0.61667   (44.21667, 0.61667)     1814.0   
8  1930-01-01 -31.60000  -65.23333    (-31.6, -65.23333)     1930.0   
9  1920-01-01 -30.86667  -64.55000   (-30.86667, -64.55)     1920.0   
10 1974-01-01  16.39806   -9.57028  (16.39806, -9.57028)     1974.0   
11 1925-01-01  19.08333    8.38333   (19.08333, 8.38333)     1925.0   
12 1769-01-01  50.66667    2.33333   (50.66667, 2.33333)     1769.0   
13 1949-01-01  29.51667   35.05000     (29.51667, 35.05)     1949.0   
14 1838-01-01  29.71667   77.95000     (29.71667, 77.95)     1838.0   
15 1959-01-01   8.91667    8.43333    (8.91667, 8.43333)     1959.0   

   Year above or equel 1970?  
0                       True  
1                       True  
2                       True  
3                      False  
4                       True  
5                       True  
6                       True  
7                       True  
8                       True  
9                       True  
10                     False  
11                      True  
12                      True  
13                      True  
14                      True  
15                      True

In [207]:
data = pd.read_csv("../data/Meteorite_Landings.csv", index_col ="id")
meteorites.sort_index(inplace=True)
meteorites.loc[10036:10040]


name    id nametype    recclass  mass (g)   fall  \
10036  Elephant Moraine 90022  8432    Valid         CK5      15.5  Found   
10037  Elephant Moraine 90023  8433    Valid         CK5      31.5  Found   
10038  Elephant Moraine 90024  8434    Valid  Eucrite-br      22.8  Found   
10039  Elephant Moraine 90025  8435    Valid         CK5      45.8  Found   
10040  Elephant Moraine 90026  8436    Valid         CK5      61.5  Found   

                         year    reclat    reclong             GeoLocation  
10036  01/01/1990 12:00:00 AM -76.28573  156.45721  (-76.28573, 156.45721)  
10037  01/01/1990 12:00:00 AM -76.27507  156.41038  (-76.27507, 156.41038)  
10038  01/01/1990 12:00:00 AM -76.28843  156.47872  (-76.28843, 156.47872)  
10039  01/01/1990 12:00:00 AM -76.28200  156.39926    (-76.282, 156.39926)  
10040  01/01/1990 12:00:00 AM -76.29226  156.45353  (-76.29226, 156.45353)

In [222]:
###### **Bonus**:
# meteorites[meteorites['year'].isna()]

# meteorites['year'] = pd.to_datetime(meteorites['year'],errors = 'coerce') 
# meteorites.loc[meteorites['year'] == 1970, 'Year above or equel 1970?'] = 'True'
# meteorites[meteorites['year'].isna()]
meteorites = pd.read_csv("../data/Meteorite_Landings.csv")

meteorites = pd.DataFrame({'year' : 'year' })
meteorites['correct'] = pd.to_datetime(meteorites['year'],errors='coerce')
print(meteorites)
# meteorites



ValueError: If using all scalar values, you must pass an index

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

In [258]:
meteorites = pd.read_csv("../data/Meteorite_Landings.csv")
meteorites['year'] = pd.to_datetime(meteorites['year'],errors='coerce')
meteorites['only_year'] = pd.DatetimeIndex(meteorites['year']).year
meteorites.rename(columns={'mass (g)':"mass"}, inplace=True)

meteorites.head()

name   id nametype     recclass      mass  fall       year    reclat  \
0    Aachen    1    Valid           L5      21.0  Fell 1880-01-01  50.77500   
1    Aarhus    2    Valid           H6     720.0  Fell 1951-01-01  56.18333   
2      Abee    6    Valid          EH4  107000.0  Fell 1952-01-01  54.21667   
3  Acapulco   10    Valid  Acapulcoite    1914.0  Fell 1976-01-01  16.88333   
4   Achiras  370    Valid           L6     780.0  Fell 1902-01-01 -33.16667   

     reclong           GeoLocation  only_year  
0    6.08333     (50.775, 6.08333)     1880.0  
1   10.23333  (56.18333, 10.23333)     1951.0  
2 -113.00000    (54.21667, -113.0)     1952.0  
3  -99.90000     (16.88333, -99.9)     1976.0  
4  -64.95000   (-33.16667, -64.95)     1902.0

In [260]:
fell_data = meteorites.query("fall == 'Fell'" & (only_year.loc[2005:2009]))
fell_data

NameError: name 'only_year' is not defined

In [240]:
meteorites['mass'].quantile([0.95])

0.95    4000.0
Name: mass, dtype: float64

In [233]:
meteorites.fall.value_counts()

Found    44609
Fell      1107
Name: fall, dtype: int64

In [248]:
sum1 = meteorites[meteorites.fall == 'Found'].sum()['mass']
sum1

/var/folders/d2/lk9wxmfn7l1b8l10pld39yxw0000gn/T/ipykernel_31434/3764445612.py:1: FutureWarning: The default value of numeric_only in DataFrame.sum is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  sum1 = meteorites[meteorites.fall == 'Found'].sum()['mass']


554680191.9879999

In [253]:
sum2 = meteorites[meteorites.fall == 'Fell'].sum()['mass']
print(sum2)
sum_both = sum1 + sum2
print(sum_both)

50601018.64999999
605281210.6379999


/var/folders/d2/lk9wxmfn7l1b8l10pld39yxw0000gn/T/ipykernel_31434/568382897.py:1: FutureWarning: The default value of numeric_only in DataFrame.sum is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  sum2 = meteorites[meteorites.fall == 'Fell'].sum()['mass']


In [255]:
meteorites(sum_both).quantile([0.95])

TypeError: 'DataFrame' object is not callable

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.